### DISCUSSION TITLE
Looking for help setting up

### DISCUSSION THREAD

**Discussion by lmieulet** [privatenumber/instant-mocha#10] [Help wanted]
Hello,

I struggle to use your fork with Webpack 5:

```
node:14344) [DEP_WEBPACK_MODULE_ID] DeprecationWarning: Module.id: Use new ChunkGraph API
Error: Module.id: There was no ChunkGraph assigned to the Module for backward-compat (Use the new API)
    at /home/leo/CS/rs-dev/regards-oss-frontend/webapp/plugins/criterion/data-with-picture-only/node_modules/webpack/lib/ChunkGraph.js:1548:12
    at deprecated (internal/util.js:89:15)
    at Function.getChunkGraphForModule (/home/leo/CS/rs-dev/regards-oss-frontend/webapp/plugins/criterion/data-with-picture-only/node_modules/webpack/lib/ChunkGraph.js:1539:18)
    at CompatGetDefaultExportRuntimeModule.get id [as id] (/home/leo/CS/rs-dev/regards-oss-frontend/webapp/plugins/criterion/data-with-picture-only/node_modules/webpack/lib/Module.js:172:21)
    at /home/leo/CS/rs-dev/regards-oss-frontend/webapp/plugins/criterion/data-with-picture-only/node_modules/mochapack/src/webpack/util/getAffectedModuleIds.ts:43:51
    at Array.reduce (<anonymous>)
    at Set.set.<computed> [as reduce] (/home/leo/CS/rs-dev/regards-oss-frontend/webapp/plugins/criterion/data-with-picture-only/node_modules/webpack/lib/util/deprecation.js:87:35)
    at buildModuleMap (/home/leo/CS/rs-dev/regards-oss-frontend/webapp/plugins/criterion/data-with-picture-only/node_modules/mochapack/src/webpack/util/getAffectedModuleIds.ts:42:29)
    at Object.getAffectedModuleIds [as default] (/home/leo/CS/rs-dev/regards-oss-frontend/webapp/plugins/criterion/data-with-picture-only/node_modules/mochapack/src/webpack/util/getAffectedModuleIds.ts:144:32)
    at Object.getBuildStats [as default] (/home/leo/CS/rs-dev/regards-oss-frontend/webapp/plugins/criterion/data-with-picture-only/node_modules/mochapack/src/webpack/util/getBuildStats.ts:20:47)
```

did you test with a real project ?

**Comment by lmieulet** [comment]
You can reproduce using :

```
git clone https://github.com/serut/regards-frontend-webpack-5
cd webapp
npm i
npm run test:mocha
```

Webpack config is inside `/webapp/webpack-config-front/src/app/webpack.test.config.js`

**Comment by privatenumber** [comment]
Yep, tested and working with a large enterprise project.

Thanks for the repro. I'm getting a different error than you. Can you provide your `instant-mocha` command?

Could it be that even though you [installed `instant-mocha`](https://github.com/serut/regards-frontend-webpack-5/blob/456c6b95f0f0e36739b7946d480af9351d1b2234/webapp/package.json#L240), you're still using [`mochapack`](https://github.com/serut/regards-frontend-webpack-5/blob/456c6b95f0f0e36739b7946d480af9351d1b2234/webapp/package.json#L37)?

PS, this isn't a fork

**Comment by privatenumber** [comment]
Also seems like you [upgraded to Webpack 5](https://github.com/serut/regards-frontend-webpack-5/commit/d55de663870281bc2d0e99b0965b014a3202e537#diff-006cce2fd88392a0bff4cf16091d480b6238c6bdf52c9042bf7805879c791a8f) without confirming that `instant-mocha` is working. So the error you provided above is completely irrelevant to `instant-mocha`.

`instant-mocha` supports both Webpack 4 & 5. Please test in isolation with Webpack 4 if that was already working for you.

And if possible, please provide a minimal reproduction of the error.

**Comment by lmieulet** [comment]
With or without webpack 5 I have the same error (after fixing test cmd) : 
```
Cannot find module 'chai'
Require stack:
- /main.js
- /home/leo/CS/rs-dev/regards-oss-frontend/webapp/node_modules/instant-mocha/dist/lib/memfs.js
- /home/leo/CS/rs-dev/regards-oss-frontend/webapp/node_modules/instant-mocha/dist/lib/mocha.js
- /home/leo/CS/rs-dev/regards-oss-frontend/webapp/node_modules/instant-mocha/dist/instant-mocha.js
- /home/leo/CS/rs-dev/regards-oss-frontend/webapp/node_modules/instant-mocha/dist/cli.js
- /home/leo/CS/rs-dev/regards-oss-frontend/webapp/node_modules/instant-mocha/bin/instant-mocha.js
npm ERR! code ELIFECYCLE
```

Sorry I thought it was a fork..

**Reply by privatenumber** [reply]
This should be fixed with v1.2.0 released yesterday via https://github.com/privatenumber/instant-mocha/pull/8.

**Comment by lmieulet** [comment]
If I had `/chai/,` to the whitelist inside `externals: [nodeExternals` it founds chai, but then it's another dependency that's missing...

**Comment by lmieulet** [comment]
I removed `externals` and `module.noParse` from my webpack test conf to make it compiles. Now some mocks are not working anymore, but it looks like a lot of things are working as expected.
Congratz !

**Comment by lmieulet** [comment]
Now I've fixed my stub, the runner does not stop when finished (I'm still using Webpack 4).

The command just hangs with : 

```
  3082 passing (3s)

[mochawesome] Report JSON saved to /home/leo/CS/rs-dev/regards-oss-frontend/webapp/reports/mocha/mochawesome.json

[mochawesome] Report HTML saved to /home/leo/CS/rs-dev/regards-oss-frontend/webapp/reports/mocha/mochawesome.html


```


Edit: Same behavior with Webpack 5.
I pushed a new branch feature/webpack5 : https://github.com/serut/regards-frontend-webpack-5/tree/feature/webpack5

**Comment by privatenumber** [comment]
Moved this to Discussions "Help wanted" because there isn't a clear bug to tackle in this thread.

Let me know if you can provide a minimal reproduction repo and I can look into it for you.

I can look into adding mochawesome to the tests.

**Comment by lmieulet** [comment]
That's all fine. I don't know if that's usual from mochapack or the original package to get the configuration I had, I just needed to remove some config keys to get it working, then appears some new babel issues due to the fresh transpilation you do. Indeed, there is the nodeExternals in their doc you must not use with your lib : https://sysgears.github.io/mochapack/docs/installation/webpack-configuration.html

I'm still strungling with the CLI not stopping after test. If I put few tests files, that's ok, but with all files I just hang out. 

Do you receive an event from Mocha when tests are over ? How does it work ? 

I'm doing some tests (restricting the CLI to tests less amount of files) and in a few minuts I will know if there is a specific test that blocks the CLI.

**Reply by privatenumber** [reply]
It's recommended: [instant-mocha#are-node_modules-automatically-externalized](https://github.com/privatenumber/instant-mocha#are-node_modules-automatically-externalized).

**Reply by lmieulet** [reply]
I can try again, I may be wrong...

**Reply by privatenumber** [reply]
Make sure you're using v1.2.0 which has the fix via https://github.com/privatenumber/instant-mocha/pull/8

You might've missed [my comment above](https://github.com/privatenumber/instant-mocha/discussions/10#discussioncomment-671458).

**Reply by lmieulet** [reply]
Not an expert but it looks like it breaks my resolve.alias configuration and bypass it. I put a stub of some lib using resolve.alias and when I put some nodeExternals it does not work anymore

**Reply by privatenumber** [reply]
That's expected behavior. `webpack-node-externals` externalizes (or removes) certain modules from the Webpack build which means the aliases won't be resolved in those modules.

You can resolve this issue by configuring `webpack-node-externals` to pass in an allowlist of modules you want to keep bundled in.

(`instant-mocha` doesn't touch `resolve.alias`)

**Comment by lmieulet** [comment]
Ok I found the buggy test inside a folder. So nothing to change on your package :tada: 
Just your doc (something like `Migration from mochapack`) :eyes:

**Reply by lmieulet** [reply]
Okey I need to unmount a component manually to remove a setTimeout (that componentDidUnmount would call). Otherwise the test was blocking the CLI to exit...

**Reply by privatenumber** [reply]
Is there anything specific you'd like me to document?

Your first issue [here](https://github.com/privatenumber/instant-mocha/discussions/10#discussion-3342499) is unrelated to `instant-mocha` and seems like a Webpack 5 issue.

Your second issue [here](https://github.com/privatenumber/instant-mocha/discussions/10#discussioncomment-671453) is resolved with v1.2.0 via https://github.com/privatenumber/instant-mocha/pull/8. You should be able to revert any changes you made to try to fix it.

Your third issue seems to be a bug in your tests.

**Reply by lmieulet** [reply]
Yeah the first issue was because I didnt use your cli

Second issue : `nodeExternals` is working perfectly  
From 1m23,689s to 0m59,122s

Third issue: the way you launch test has an impact on setTimeout when used in react component. I don't know why..

**Reply by privatenumber** [reply]
If there's anything specific you want me to document, let me know.

**Reply by lmieulet** [reply]
maybe coverage with nyc and a sample repository to show how it looks like - for example the inclusion of nodeExternals is adviced in the real word.

**Comment by lmieulet** [comment]
Hello @privatenumber 
Did you tryed the test watch mode ? I tried it, the first execution works well, but then when I make some file edition, it just run tests again and again and it never stops

**Reply by privatenumber** [reply]
Works fine for me

**Reply by lmieulet** [reply]
Ok so there is something wrong inside my 3K tests ...

**Reply by privatenumber** [reply]
I'd like to help but I can't unless I see a minimal reproduction. If you can create one, let me know and I'll check it out.

**Reply by lmieulet** [reply]
if you don't have this behavior, it must be related to my test files or my webpack plugin. I will look that and tell you
